### DISCUSSION TITLE
Question about DPatch implementation

### DISCUSSION THREAD

**Discussion by yuki2222** [Trusted-AI/adversarial-robustness-toolbox#1773] [Q&A] [answered]
Hi,

I have some question regarding the DPatch code on here: `/adversarial-robustness-toolbox/tree/main/art/attacks/evasion)/dpatch.py`.

From the line 163 through 201, it goes:

        if self.target_label:

            for i_image in range(patched_images.shape[0]):
                if isinstance(self.target_label, int):
                    t_l = self.target_label
                else:
                    t_l = self.target_label[i_image]

                i_x_1 = transforms[i_image]["i_x_1"]
                i_x_2 = transforms[i_image]["i_x_2"]
                i_y_1 = transforms[i_image]["i_y_1"]
                i_y_2 = transforms[i_image]["i_y_2"]

                target_dict = {}
                target_dict["boxes"] = np.asarray([[i_x_1, i_y_1, i_x_2, i_y_2]])
                target_dict["labels"] = np.asarray(
                    [
                        t_l,
                    ]
                )
                target_dict["scores"] = np.asarray(
                    [
                        1.0,
                    ]
                )

                patch_target.append(target_dict)

        else:

            predictions = self.estimator.predict(x=patched_images, standardise_output=True)

            for i_image in range(patched_images.shape[0]):
                target_dict = {}
                target_dict["boxes"] = predictions[i_image]["boxes"]
                target_dict["labels"] = predictions[i_image]["labels"]
                target_dict["scores"] = predictions[i_image]["scores"]

                patch_target.append(target_dict)

For the untargeted attack it is taking the estimator predictions as the patch target, but shouldn’t this take the true labels (annotated labels) instead? As far as I understand, the original paper says to maximize the loss of the estimator to the true labels (page 3:  https://arxiv.org/abs/1806.02299v4).

Sorry if I’m missing some point, but I am quite confused with this one. I’d really appreciate if someone could help me out. Thanks you!

**Comment by beat-buesser** [comment, marked as answer]
Hi @yuki2222 You are right, the current implementation of `DPatch` in the untargeted case attacks the object detector's predictions on the patched images and does not accept true labels. I think the reason was that the predictions to be attacked are made on the images with the untrained patch applied in an approach to automatically remove the influence of the patch on the object detection accuracy under the assumption that the predicted and true labels are similar.
I think it would be easily possible to add support for true labels, would you need it for an application or to exactly reproduce `DPatch`?

**Reply by yuki2222** [reply]
Thank you for the response @beat-buesser ! That makes sense. I need to use `DPatch` to experiment its effect on the object detectors used in the self-driving cars, as part of my work project. So it doesn’t have to be exactly as in the paper, as long as I could do the experiment :)

Another question I had was that when `DPatch` attacks the detector’s predictions in the untargeted case, the computed loss could technically be 0.0 since it is comparing the new predictions to the said predictions, both of which are made on the same patched image by the same detector. But as you answered, the predictions to be attacked are made without the influence of the patch in your implementation, thus those two predictions are not exactly the same. Is my understanding correct?

**Reply by beat-buesser** [reply]
Hi @yuki2222 That's a great question about the loss! I think we have not noticed it so far because we have been testing it mainly with torchvision's FasterRCNN which seems to be non-deterministic which results in small variations among consecutive forward/backward passes and non-zero loss values.

I think for experiments it would be better to have the option to provide true labels. It should only require a few lines to be added. Would you be interested to open a pull request? I would be happy to provide support.

Btw, have you already seen our implementation of `RobustDPatch`?

**Reply by yuki2222** [reply]
Hi @beat-buesser I see! I didn’t know the non-deterministic character of FasterRCNN. I’m also thinking of using FasterRCNN (and YOLO) for my experiment, so it should work. I will try it when I have time this week.

As for the pull request, I can try! This will actually be my first time to do it on GitHub. Should I make changes on the code and submit it to you? I’m just worried about if my code would be good enough for that haha..

I haven’t checked out `RobustDPatch` yet! I will take a look later. How is it different from the normal `DPatch`?

**Reply by beat-buesser** [reply]
Hi @yuki2222 Great! Your code will be perfect! We would have to make the following changes:
- remove the if-block in line 137
- update the if-condition in line 163 to include the `y` has to be `None` for targeted attacks
- extend line 193 to use `y` if it is not None, else use the estimator's predictions

To get started you would have to the following steps, we can discuss each step in more detail if needed:
- for this repo to create your own copy of it
- clone the forked repo to you machine
- switch to branch `dev_1.12.0`
- make changes to the code and commit using `git` with option `-s`
- push the commits to your forked repo
- open a pull request to the upstream repo
- we'll review and merge the PR

There was a bug in the code used for the paper (not using ART) and a follow-up paper (https://arxiv.org/abs/1906.11897) fixed the bug and increased the attack strength with additional extension including expectation over transformation. We have implemented these upgrades in `RobustDPatch`.

**Reply by yuki2222** [reply]
Hi @beat-buesser Thank you for the details! I am almost done with the changes. Here is the code I made: 2be96ee580f15bd6aacd64f2320cc8ccf47fe891

I changed line 193 to below. Does this look okay? I feel like this could throw an error as y may not always have the keys like "boxes" or "scores" inside the for loop, depending on the format of y.

        else:

            predictions = y if y is not None else self.estimator.predict(x=patched_images, standardise_output=True)

            for i_image in range(patched_images.shape[0]):
                target_dict = {}
                target_dict["boxes"] = predictions[i_image]["boxes"]
                target_dict["labels"] = predictions[i_image]["labels"]
                target_dict["scores"] = predictions[i_image]["scores"]

                patch_target.append(target_dict)

I had seen that paper just a bit before, but didn't know it was also implemented here. Thanks for sharing!

**Reply by beat-buesser** [reply]
Hi @yuki2222 I think your code looks good. I agree on the possibility of an error if the format of `y` is not as expected. To better inform the user on the expected format of `y`, could you please add the following change to the docstring of method `generate` of `DPatch` by replacing line 110 with:
```
        :param y: True labels of type `List[Dict[np.ndarray]]` for untargeted attack, one dictionary per input image. The
                  keys and values of the dictionary are:

                 - boxes [N, 4]: the boxes in [x1, y1, x2, y2] format, with 0 <= x1 < x2 <= W and 0 <= y1 < y2 <= H.
                 - labels [N]: the labels for each image
                 - scores [N]: the scores or each prediction.
```

**Reply by yuki2222** [reply]
Hi @beat-buesser Thank you! I added your comment to the code and just sent the pull request. I ran some test with it, so it should work :)

**Reply by beat-buesser** [reply]
Hi @yuki2222 That's great! Thank you very much. I'll start the review of the PR.

**Reply by beat-buesser** [reply]
Hi @yuki2222 I have merged #1780 Congratulations to your first PR to ART!
